lens/sqlrepo/tstracker/tracker.go

```diff
@@ -0,0 +1,283 @@
+package tstracker
+
+import (
+	"bytes"
+	"context"
+	"encoding/binary"
+	"time"
+
+	"github.com/jackc/pgx/v4"
+	"golang.org/x/xerrors"
+
+	"github.com/filecoin-project/go-address"
+	pgchainbs "github.com/filecoin-project/go-bs-postgres-chainnotated"
+	"github.com/filecoin-project/go-state-types/abi"
+	"github.com/filecoin-project/lotus/chain/types"
+	"github.com/ipfs/go-cid"
+)
+
+func (tcs *tcs) CurrentDbTipSetKey(ctx context.Context) (*types.TipSetKey, abi.ChainEpoch, error) {
+
+	tsks, epoch, err := tcs.CurrentFilTipSetKey(ctx)
+	if err != nil {
+		return nil, -1, err
+	}
+
+	tsk := types.NewTipSetKey(tsks...)
+
+	return &tsk, epoch, nil
+}
+
+func (tcs *tcs) GetCurrentTipset(ctx context.Context, lookback int) (*types.TipSetKey, error) {
+	tsks, _, err := tcs.CurrentFilTipSetKey(ctx)
+	if err != nil {
+		return nil, err
+	}
+	tsk := types.NewTipSetKey(tsks...)
+
+	return &tsk, nil
+}
+
+func (tcs *tcs) StoreTipSetVist(ctx context.Context, ts *types.TipSet, isHeadChange bool) error {
+
+	// note the time before anything else
+	visitTime := time.Now()
+
+	if !tcs.IsWritable() {
+		return xerrors.New("unable to StoreTipSetVist on read-only chainstore")
+	}
+	if tcs.InstanceNamespace() == "" {
+		return xerrors.New("unable to StoreTipSetVist without a specified instance namespace")
+	}
+
+	headers := ts.Blocks()
+	// somehow got called with a null round - nothing to record
+	if len(headers) == 0 {
+		return nil
+	}
+
+	// if a headchange: take an early lock - this allows us to properly serialize
+	// events from competing workers
+	if isHeadChange {
+		tcs.lastSeenTipSetMu.Lock()
+		defer tcs.lastSeenTipSetMu.Unlock()
+
+	} else {
+		tcs.lastSeenTipSetMu.RLock()
+		lastTs := tcs.lastSeenTipSet
+		tcs.lastSeenTipSetMu.RUnlock()
+
+		// make no records for duplicate visits ( they do happen inexplicably )
+		if lastTs != nil && lastTs.Equals(ts) {
+			return nil
+		}
+	}
+
+	tskCidsBytes := make([][]byte, 0, len(headers))
+	for _, h := range headers {
+		tskCidsBytes = append(tskCidsBytes, h.Cid().Bytes())
+	}
+
+	// check for the TS record: can skip destructuring (which isn't cheap)
+	var tsDbOrdinal *int32
+	err := tcs.PgxPool().QueryRow(
+		ctx,
+		`SELECT tipset_ordinal FROM fil_common_base.tipsets WHERE tipset_cids = $1::BYTEA[]`,
+		tskCidsBytes,
+	).Scan(&tsDbOrdinal)
```

Comment:
Is this correct? Passing pointer to pointer?

Reply:
In this very specific case likely not. But in general pgx uses double pointers like this to handle NULL returns. Because the return of this function is a pointer in itself, it ended up being used as per pgx regime.